### PR TITLE
Eqy check

### PR DESCRIPTION
**Pull request by openroadie**
Adding eqy checks into ORFS with help from the eqy team on how to structure the input config file.

### CODE REVIEW

**Comment by openroadie** [comment]
@vvbandeira This PR cannot be merged until the CI/CD container has these repos  installed as the default (sudo make install). 
Please follow the order the repos are mentioned in:

https://github.com/YosysHQ/yosys
https://github.com/YosysHQ/eqy
https://github.com/YosysHQ/sby

Please do let me know if I can help in any way. Thx !

File: flow/scripts/load.tcl [hunk 1]
```diff
@@ -33,3 +33,59 @@ proc load_design {design_file sdc_file msg} {
...
+proc get_verilog_cells_for_design { } {
+    set dir "$::env(PLATFORM_DIR)/work_around_yosys/"
+    set cell_files [glob $dir/*]
+}
+
+proc write_eqy_golden_verilog {} {
+    write_verilog before.v
+    # Filter out tap cells. 
+    exec cat before.v | grep -v TAPCELL > before.1.v
+    exec mv before.1.v before.v
```
Comment (maliberty) [inline flow/scripts/load.tcl]: Could -remove_cells be used for this?

File: flow/scripts/load.tcl [hunk 2]
```diff
@@ -33,3 +33,59 @@ proc load_design {design_file sdc_file msg} {
...
+#===========================================================================================
+# Routines to run equivalence tests when they are enabled. 
+
+proc get_verilog_cells_for_design { } {
+    set dir "$::env(PLATFORM_DIR)/work_around_yosys/"
+    set cell_files [glob $dir/*]
+}
+
+proc write_eqy_golden_verilog {} {
+    write_verilog before.v
```
Comment (maliberty) [inline flow/scripts/load.tcl]: This should be written into the results dir not the current dir to avoid collisions.

File: flow/scripts/load.tcl [hunk 3]
```diff
@@ -33,3 +33,59 @@ proc load_design {design_file sdc_file msg} {
...
+}
+
+proc run_equivalence_test {} {
+    write_verilog after.v
+    exec cat after.v | grep -v TAPCELL > after.1.v
+    exec mv after.1.v after.v
+    write_eqy_script
+    
+    exec rm -rf eqy_test    
+    exec eqy eqy_test.eqy > output.log
```
Comment (maliberty) [inline flow/scripts/load.tcl]: write to log dir

File: flow/scripts/load.tcl [hunk 4]
```diff
@@ -33,3 +33,59 @@ proc load_design {design_file sdc_file msg} {
...
+    set count [exec grep -c "Successfully proved designs equivalent" output.log]
+    if { $count == 0 } {
+	puts "Repair timing output failed equivalence test"
```
Comment (maliberty) [inline flow/scripts/load.tcl]: This should fail the run otherwise this message will just get lost in log file.

File: flow/scripts/load.tcl (hunk 1 above)
Comment (openroadie) [reply flow/scripts/load.tcl]: Good idea. Let me try.

File: flow/scripts/load.tcl [hunk 5]
```diff
@@ -33,3 +33,59 @@ proc load_design {design_file sdc_file msg} {
...
+    write_verilog before.v
+    # Filter out tap cells. 
+    exec cat before.v | grep -v TAPCELL > before.1.v
+    exec mv before.1.v before.v
+}
+
+proc write_eqy_script { } {
+    set top_cell [current_design]
+    set cell_files [get_verilog_cells_for_design]
+    set outfile [open "eqy_test.eqy" w]
```
Comment (maliberty) [inline flow/scripts/load.tcl]: into objects

Comment (openroadie) [reply flow/scripts/load.tcl]: Unclear what "into objects" means.

Comment (maliberty) [reply flow/scripts/load.tcl]: $(OBJECTS_DIR)

**Comment by vijayank88** [comment]
@openroadie 
Is that doing logic equivalence check?
Also plz update `EQUIVALENCE_CHECK ` variable to part of [FlowVariables.md](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/docs/user/FlowVariables.md) with descriptions.

File: flow/scripts/load.tcl [hunk 6]
```diff
@@ -79,14 +79,14 @@ proc run_equivalence_test {} {
     write_eqy_script
 
     if { [file exists  $::env(LOG_DIR)/4_eqy_output] } {
-	exec "/bin/rm -rf $::env(LOG_DIR)/4_eqy_output"
+      eval exec rm -rf $::env(LOG_DIR)/4_eqy_output
```
Comment (rovinski) [inline flow/scripts/load.tcl]: Use `file delete -force $::env(LOG_DIR)/4_eqy_output`
No need to check for existence either. `file delete` doesn't error on nonexistent paths.

File: flow/scripts/cts.tcl [hunk 7]
```diff
@@ -98,8 +98,17 @@ if { [info exists ::env(SKIP_GATE_CLONING)] } {
   append additional_args " -skip_gate_cloning"
 }
 
+
+if { [info exists ::env(EQUIVALENCE_CHECK)] } {
+    write_eqy_verilog 4_before_rsz.v
```
Comment (vvbandeira) [inline flow/scripts/cts.tcl:103]: Should this be written in the `objects` folder?

**Review by vvbandeira — COMMENTED** [review]
Just this small question. CI is in a shape where is good to merge the PR, please let me know when to merge -- I would like to do a "squash and merge" due to the number of commits.

Edit: maybe add an exclude to the `4_eqy_output` folder so as to not save in CI as it contains a lot of files which is slowing down the pipeline?

File: flow/scripts/cts.tcl (hunk 7 above)
Comment (openroadie) [reply flow/scripts/cts.tcl:103]: lets merge this PR for now and deal with this later. I do not have a strong opinion on where this files goes.

**Comment by openroadie** [comment]
@vvbandeira this is good to merge from my point of view. But CI/CD is still failing (pr-head) as far as I can see.

**Comment by huang-kaixin** [comment]
> Adding eqy checks into ORFS with help from the eqy team on how to structure the input config file.
May I ask why [adding eqy checks](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/7adca25d4ec5604fea194a695c62b2575db65809/flow/designs/sky130hd/gcd/config.mk#L12C1-L12C34) is necessary? I encountered many "110 could not execute" eqy "errors while running gcd.

**Comment by maliberty** [comment]
eqy is trying to prove that OR hasn't introduced any logic errors in the course of optimization.  We did hit one such case a while ago and wanted to have some checking.

**Comment by refaay** [comment]
I am trying to build and execute a fresh clone of ORFS.
I am also facing the same issue "Error: cts.tcl, 113 couldn't execute "eqy": no such file or directory" while running "make" in "OpenROAD-flow-scripts/flow" and then it breaks.
How do I fix it?

**Comment by maliberty** [comment]
@vvbandeira please help @refaay with eqy

**Pull request merged**